Exposing new TRT build options

gedoensmax opened the pull request:
### Description

This will add a few TRT options, some of them are only available on TRT 8.6:
- heuristics
- sparsity
- optimization level (8.6 only)
- auxiliary stream (8.6 only)
- tactic source selection

I am no sure yet which tests is should add for these options. As those are mostly simple TRT flags i am not sure to what level i should test. For heuristics something similar to https://github.com/gedoensmax/onnxruntime/blob/44dda08b51e98bd86f6eafd09b7d00c3680c78a8/onnxruntime/test/providers/tensorrt/tensorrt_basic_test.cc#L510-L538 should be possible for, but for all other essentially we would only be testing if there is a crash or not if the option is set.
Also if i forgot some option that would be good to have feel free to speak up !

sl1pkn07 commented:
Hi. is possible backport to 1.14?

greetings

gedoensmax commented:
@chilo-ms can you comment on the granularity of tests to add ?

chilo-ms commented:
> @chilo-ms can you comment on the granularity of tests to add ?

If there are some obvious outputs tied to specific TRT flags, we should test the existence of them. Otherwise, testing inference and verify simple input/output should be enough.

gedoensmax commented:
@jywu-msft I added unit tests, rebased on main and added a compile time decision to also support older TRT versions. Please give this a fresh review and let me know if there is something else.

jywu-msft commented:
> @jywu-msft I added unit tests, rebased on main and added a compile time decision to also support older TRT versions. Please give this a fresh review and let me know if there is something else.

Thanks! We'll take a look.

jywu-msft commented:
/azp run Linux GPU TensorRT CI Pipeline, Windows GPU TensorRT CI Pipeline

chilo-ms commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, ONNX Runtime Web CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, Linux QNN CI Pipeline

chilo-ms commented:
/azp run Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows ARM64 QNN CI Pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed, ONNX Runtime React Native CI Pipeline

chilo-ms commented:
Could you also help add those provider options to onnxruntime_perf_test in `onnxruntime/test/perftest/ort_test_session.cc`?
Just like you did for the timing cache [here](https://github.com/microsoft/onnxruntime/pull/14767/files#diff-f3908cf4d2982fd2b008beb8c951d9cab67b28cb07077493b8d1dd8b448d6e18). Thank you.

gedoensmax commented:
Thanks for noting ! I added the options.

chilo-ms commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, ONNX Runtime Web CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, Linux QNN CI Pipeline

chilo-ms commented:
/azp run Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows ARM64 QNN CI Pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed, ONNX Runtime React Native CI Pipeline

chilo-ms commented:
/azp run Python format

chilo-ms commented on onnxruntime/test/shared_lib/test_inference.cc:
```diff
@@ -2508,11 +2515,13 @@ TEST(CApiTest, TestConfigureTensorRTProviderOptions) {
+                                  engine_cache_path, option_value.c_str()};
 
-  ASSERT_TRUE(api.UpdateTensorRTProviderOptions(rel_trt_options.get(), keys.data(), values.data(), 5) == nullptr);
+  ASSERT_TRUE(api.UpdateTensorRTProviderOptions(rel_trt_options.get(), keys.data(), values.data(), keys.size()) == nullptr);
```
Here is a minor suggestion. 
Once you call api.UpdateTensorRTProviderOptions(), you might want to check the whether the provider options are being set simply through checking the returnin string from api.GetTensorRTProviderOptionsAsString().

You can reference the code `ASSERT_TRUE(s.find(engine_cache_path) != std::string::npos)` below and doing the same for trt_tactic_sources=-CUDNN,+CUBLAS ?

gedoensmax replied:
Also while writing that test i discovered some typo's on my end so it was a good idea to do it 👍

chilo-ms commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, ONNX Runtime Web CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, Linux QNN CI Pipeline

chilo-ms commented:
/azp run Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows ARM64 QNN CI Pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed, ONNX Runtime React Native CI Pipeline

chilo-ms commented:
You might need to merge main. 
The dml pipelines failed and I think the latest [PR](https://github.com/microsoft/onnxruntime/pull/15468) addressed this issue.
Other than that, I think we are good to go.

gedoensmax commented:
@chi-lo I just rebased the branch - we should be good now.

chilo-ms commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, ONNX Runtime Web CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline, Linux QNN CI Pipeline

chilo-ms commented:
/azp run Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows ARM64 QNN CI Pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed, ONNX Runtime React Native CI Pipeline

chilo-ms commented:
> Any recommendations on which commit will solve the failing CI ? My impression is that this is unrelated to my changes.
The failing CIs are not required to get this merged. We can merge now.

chilo-ms commented:
The failing CIs are not required to be fixed to get this PR merged. We can merge it now.
